Fix #140: Minimize redundant logical component details

lgolding opened the pull request:
Also:
- Revise values for `logicalLocationComponent.kind`.
- Specify that both `file.pathFromParent` and
  `logicalLocationComponent.name` can be optional in
  certain cases, and give examples.

@michaelcfanning

Commit:
Fix #140: Minimize redundant logical component details

Also:
* Revise values for `logicalLocationComponent.kind`.
* Specify that both `file.pathFromParent` and
    `logicalLocationComponent.name` can be optional in
    certain cases, and give examples.

lgolding commented on document/CoreFormat.mdk:
```diff
@@ -508,6 +552,43 @@ whose value is a string specifying that path.
 If the location of the file within its containing property cannot be expressed by means of a path,
 then the `pathFromParent` property shall be absent.
 
+If the `file` object occurs as the only member of an array within the `run.files` object (&sect;[#run-files]),
```
You can't tell from GitHub's context, but this text is added to the section on `file.pathFromParent`.

lgolding commented on document/CoreFormat.mdk:
```diff
@@ -1087,28 +1167,13 @@ that identifies the file where the static analysis tool produced the result.
 Whether `resultFile` is present depends on the information available to the tool that produces the log file
 (see &sect;[#location-Constraints]).
 
-### `logicalLocation` property { #location-logicalLocation }
```
Most of this text moved up into `run.logicalLocations` (see above, in the right-hand diff).

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -340,6 +340,48 @@ run: {
+      "kind": "namespace"
+    },
+    {
+      "name": "namespaceB",
```
Can we remove the double-nesting of namespace here? this just seems to unnecessarily encourage bloating the log file,. i.e., doesn't buy us anything from simply expressing the namespace as a single item.

lgolding replied:
As we discussed: Will keep as is so consumers don't have to parse when rolling up results within a namespace and its sub-namespaces.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -508,6 +552,43 @@ whose value is a string specifying that path.
 then the `pathFromParent` property shall be absent.
 
+If the `file` object occurs as the only member of an array within the `run.files` object (&sect;[#run-files]),
+then the `pathFromParent` property may be absent.
```
we're renaming this to 'uri', yes??

lgolding replied:
Ugh. Yes.

lgolding replied:
Are you making the SDK change?

michaelcfanning replied:
yes

---

In reply to: [60333407](https://github.com/sarif-standard/sarif-spec/pull/161#discussion_r60333407) [](ancestors = 60333407)

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -508,6 +552,43 @@ whose value is a string specifying that path.
+If it is present, it must be equal to the corresponding property name within the `run.files` object,
+
+~ IsoNote
+In this case, it might be interpreted as a path from a "null" parent.
```
And then you can drop this case, since an absolute uri stands on its own without explanation...

lgolding replied:
Done.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -996,7 +1077,10 @@ This allows tools to include information about the result that is not explicitly
-and it may also specify the logical location (&sect;[#location-logicalLocation]) of the result.
+and it may also specify the logical location (&sect;[#location-fullyQualifiedLogicalName]) of the result.
+
+A logical location specifies a programmatic construct, for example, a class name or a function name,
```
this particular sentence isn't obviously illuminating.

lgolding replied:
It's been there all along; I just moved it up here from further down. I think it says something useful ("don't include a file name in this thing!"). I'm open to suggestions for rephrasing it.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -1087,28 +1167,13 @@ that identifies the file where the static analysis tool produced the result.
-in a format consistent with the programming language in which the programming constructs specified by that array were expressed.
+which specifies the fully qualified name of the logical location where the analysis tool produced the result.
+
+The format of the `fullyQualifiedLogicalName` string shall be consistent with the programming language
```
The fullyQualifiedLogicalName is an important contributor to fingerprinting scenarios, because this item is more resilient to changes in source code (whereas line locations are very fragile). This is one rationale for emitting one (and a rationale to keep the formatting consistent). In other cases, source locations may not be available (as in the absence of PDBs), in which case the logical location can help a viewer, in another context,  to actually locate the code associated with the result. Good things to mention about why this content is specified.

lgolding replied:
Added.

lgolding replied:
Added.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -1120,11 +1185,78 @@ C++: `Namespace::Class::Method(int, double) const &&`
+with one exception, described in &sect;[#location-logicalLocationKey].
+
 ~ IsoNote
 The intent of this property is to allow a result log viewer to display the logical location
```
See two other reasons I provided.

lgolding replied:
Yes. I now have bits of this in two different locations. Last change before I merge is to figure out where to combine this information and put it in one place.

lgolding replied:
Ok, I've thought about this some more. Those three rationales (1. fingerprinting, 2. analysis without PDB information, and 3. human readability) are really addressing two different issues:

1. Why include logical location information in SARIF at all, when you already have physical location information?

2. Why include the `fullyQualifiedLogicalName` property, when all the information in it is captured in the array of `logicLocationComponent` objects?

The sentence you commented on -- and remember, it was written at a time when `fullyQualifiedLogicalName` and the `logicalLocationComponent` array were adjacent to each other, both children of `location` -- was intended to answer #2, "why have `fullyQualifiedLogicalName` when you already have the more detailed `logicalLocationComponent`?" The answer was "to help viewers provide the logical information in a readable way".

On the other hand, the two rationales you cited address #1, "why have both physical _and_ logical locations?". And the answers are "because logical information is more resilient, hence better for fingerprinting", and "because it's sometimes available when source information is not".

(The argument from resilience justifies logical location information in general, not the fQLN property in particular. You could just as well fingerprint the `logicalLocationComponent` array.)

Therefore I think it's appropriate to split those two concerns in the spec. I'll address #1, with your two rationales, in the "location - General" section. I'll leave the answer to #2 where it is, in the "location - fullyQualifiedLogicalName" section. But I'll be clearer about which of the two concerns each one is addressing.

michaelcfanning replied:
Sometimes you don't have physical location data, end of story. Logical location details can help fill the gap in those cases. fullyQualifiedLogicalName exists because we want producers to create it, not consumers (if you require consumers to do it, they will make their own formatting/delimiter decisions). This property is intended for display and suppression (and it helps humans with navigation because they recognize the item in question without special parsing). The additional logical location components assist consumers in organizing data and can help viewers and other non-human processors in potentially navigating to items (where otherwise, they'd have to parse a fullyQualifiedLogicalName correctly).

---

In reply to: [60402106](https://github.com/sarif-standard/sarif-spec/pull/161#discussion_r60402106) [](ancestors = 60402106)

lgolding replied:
Much but not all of this is already in the spec:

What's there:

* Logical location fills the gap when physical location is missing. I say that in NOTE 2 in Section Location-General.

* Intended for display. I say that in the NOTE in Section Location-FullyQualifiedLogicalName ("allows a result log viewer to display the location in a way that is easily understood by users")

* Helps viewers and humans to navigate. Also mentioned in the FQLN section, but I'll add "and other programs" to the now-existing text "code viewers and end users"

Here's what's not there:

* Intended for suppression. I'll add that.

* Relieves consumers of having to format it. I'll add that.

* Explaining why we have the array of logical location components (assist in organizing). I'll add that.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -1432,12 +1566,61 @@ A C++ analysis tool might emit the `name` property of a function as the "decorat
+"logicalLocations": {
+  "functionF": [
+    {
+      "pathFromParent": "functionF",
```
this member should be 'name' :)

lgolding replied:
Good catch, fixed.

lgolding replied:
Fixed.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -1432,12 +1566,61 @@ A C++ analysis tool might emit the `name` property of a function as the "decorat
+"logicalLocations": {
+  "functionF-0": [
+    {
+      "pathFromParent": "functionF",
```
same here 'name'.

lgolding replied:
Fixed.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -1432,12 +1566,61 @@ A C++ analysis tool might emit the `name` property of a function as the "decorat
 A `logicalLocationComponent` object should contain a property named `kind`, whose value shall be one of the following strings,
 if any of those strings accurately describes the construct identified by this object:
 
-*	`method`
```
You can add method as far as I'm concerned, since we saw it in Android Studio. :)

lgolding replied:
Done.

Commit:
Address PR feedback on #140

Also, remove most (but not all) uses of the word "static" from the spec,
since almost everything in it applies to dynamic tools as well.

lgolding commented:
Also, removed most (but not all) occurrences of the word "static".

Commit:
Added "method" to logicalLocationComponent.kind

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -340,9 +340,51 @@ run: {
+a string representing the logical location where the result was produced, in a format consistent with
+the programming language in which the programmatic construct specified by that logical location was expressed.
+We refer to this string as a "fully qualified logical name".
+See &sect;[#location-fullyQualifiedLogicalName] for examples.
```
We don't refer to this string as a fully qualified name, a logical location object contains a property 'name', which is an identifier for some portion of the logical name. fullyQualifiedLogicalName lives on the location object.

lgolding replied:
The string I'm referring to here is the property name on the `logicalLocations` object ("With one exception ... each property name ... shall be a string ... in a format blah blah ... We refer to this string as a 'fully qualified logical name'."

I'm referring to the string `"namespaceA::namespaceB"` here, not to the `name` property of any of the `logicalLocationComponent` objects in the array.

```json
"logicalLocations": {
  "namespaceA::namespaceB": [
  ]
}
```

And I'm proposing to refer to a name in this form as a "fully qualified logical name", consistent with the name of the property on the `location` object.

michaelcfanning replied:
i see

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -451,15 +493,15 @@ The tool would synthesize a SemVer string `"1.1.2-beta.12"`.
 ~ IsoNumberedNote
-The rationale is that a static analysis tool knows whether its version string is intended to be
+The rationale is that an analysis tool knows whether its version string is intended to be
 interpreted according to SemVer.
```
Do we recap this guidance in the converter appendix? probably worth doing so.

lgolding replied:
I added on to Annex C.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -491,22 +533,57 @@ This allows tools to include information about the themselves that is not explic
+### `uri` property { #file-uri }
 
 If the location of a nested file with respect to its containing file can be expressed
 only by means of a path within its containing file,
```
you allow wiggle room here for producers to omit the uri, if one can be expressed but the offset could also be used. if the location can be expressed in this way, why don't we simply require it?

lgolding replied:
We could certainly have done that, but we did examine this in detail when we de-dup'd the physical location information, and this is what we decided. We can certainly revisit that decision. It's a balance between on the one hand, making life easier for producers by not requiring them to compute and emit information that the consumer doesn't strictly need, and on the other hand, making life easier for consumers by giving them as much information as possible, even if they don't need all of it.

In this particular case... I don't know. It seems that it would be a rare case where the producer knew both the offset and the uri. Your call.

michaelcfanning replied:
If you can express a URI, you should express the URI, in my opinion. There's no clear winner here: if a producer doesn't emit it, that complicates the life of a consumer. If you can provide an offset in this case as well, ok, fine, why prohibit that. It would be rare, as you say.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -994,9 +1071,27 @@ This allows tools to include information about the result that is not explicitly
+    
+2. In the absence of symbol information, binary analysis tools might not have source code locations available,
+    so information about line and column numbers might not be present in the log file.
+    In this case, code editors can use logical location to navigate from a result to the correct
```
not just code editors but human beings. absent source file details, you need to be able to find whatever an analysis result points to.

lgolding replied:
Added "or end users"

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -994,9 +1071,27 @@ This allows tools to include information about the result that is not explicitly
+There are two reasons to include logical locations in the SARIF format in addition to physical locations:
+
+1. Logical location information is an important contributor to fingerprinting scenarios,
+    because it is more resilient to changes in source code than are line locations.
```
i'd say 'tends to be more resilient' or 'it is typically more resilient'. if someone is on a namespace renaming scrub, line locations will remain constant but all your logical names may change...

lgolding replied:
Added "typically".

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -1082,33 +1173,18 @@ Whether `analysisTarget` is present depends on the information available to the
+
+The format of the `fullyQualifiedLogicalName` string shall be consistent with the programming language
+in which the programmatic construct specified by that logical location was expressed.
 
```
here's where i expected to find more useful information about the fully qualified logical name. It is a readable name that uniquely identifies the item under analysis. It is therefore highly suited to display to users to help identify the analysis target in question. As a readable item that uniquely identifies something, it is useful for fingerprinting (including in-source suppressions where, again, the readability is an advantage).

lgolding replied:
I mention this in the NOTE a couple of paragraphs down. Let me know if it's not what you had in mind.

Commit:
Additional PR feedback on #140

Commit:
Final tweaks to #140

Refine rationale for run.logicalLocations per @michaelcfanning.

michaelcfanning commented:
good stuff. +1
